### PR TITLE
WIP: edit & delete functionality (needs testing)

### PR DESCRIPTION
**Pull request by gemantzu**
Please test this and leave comments.

### CODE REVIEW

File: lib/companies/pending_changes.ex [hunk 1]
```diff
@@ -75,6 +76,18 @@ defmodule Companies.PendingChanges do
...
+  defp current_values(%{changes: %{"id" => id}, resource: "job"}) do
+    {%{id: company_id}, current_values} =
+      Job
+      |> Repo.get(id)
+      |> Repo.preload([:company])
+      |> drop_ecto_fields()
+      |> drop_nulls()
+      |> Map.pop(:company)
+
+    Map.put(current_values, :company_id, company_id)
```
Comment (doomspork) [inline lib/companies/pending_changes.ex]: If you drop the `preload` you'll be left with `company_id`, won't you?

Comment (gemantzu) [reply lib/companies/pending_changes.ex]: Damn, that's true. I guess when one does a pr @ 2 in the morning, this happens :)

File: lib/companies/jobs.ex [hunk 2]
```diff
@@ -86,8 +86,10 @@ defmodule Companies.Jobs do
       {:error, %Ecto.Changeset{}}
 
   """
-  def delete_job(%Job{} = job) do
-    Repo.delete(job)
+  def delete_job(%Job{} = job, user) do
+    job
+    |> Job.changeset(%{})
```
Comment (doomspork) [inline lib/companies/jobs.ex]: We shouldn't need to create a changeset just to delete.

Comment (gemantzu) [reply lib/companies/jobs.ex]: This is one of the things i was working on today, but didn't finish until i left home. The whole thing needs some rework, plus there are some holes we have in the original branch as well. I am going to work on them tomorrow and during the weekend.

**Comment by gemantzu** [comment]
Latest commit handles #464 as well.

File: config/dev.exs [hunk 3]
```diff
@@ -73,3 +73,5 @@ config :companies, Companies.Repo,
   database: "companies_dev",
   hostname: "localhost",
   pool_size: 10
+
+config :companies, :notify_slack, false
```
Comment (doomspork) [inline config/dev.exs]: I think we should solve this a different way.

File: lib/companies/slack.ex [hunk 4]
```diff
@@ -2,7 +2,23 @@ defmodule Companies.Slack do
   use HTTPoison.Base
 
   def notify(%{action: action, resource: resource, user: %{nickname: nickname}}) do
-    message = %{
+    case notify_slack() do
```
Comment (doomspork) [inline lib/companies/slack.ex]: I think we should configure which module to use rather than this bool. Here we lose our notifications in dev which is not useful.

```elixir
defmodule Notify.Console do
  use HTTPoison.Base

  require Logger

  def notify(type, _resource), do: Logger.info("NOTIFICATION — #{type}")
end
```

File: lib/companies/companies.ex [hunk 5]
```diff
@@ -125,8 +125,8 @@ defmodule Companies.Companies do
   {:error, %Ecto.Changeset{}}
 
   """
-  def delete_company(%Company{} = company) do
-    Repo.delete(company)
+  def delete_company(%Company{} = company, user) do
+    PendingChanges.create(%{data: company, params: %{"id" => company.id, "delete" => true}}, :delete, user)
```
Comment (doomspork) [inline lib/companies/companies.ex]: `"delete" => true` seems redundant when passing `:delete`?

File: lib/companies/jobs.ex [hunk 6]
```diff
@@ -86,8 +86,8 @@ defmodule Companies.Jobs do
       {:error, %Ecto.Changeset{}}
 
   """
-  def delete_job(%Job{} = job) do
-    Repo.delete(job)
+  def delete_job(%Job{} = job, user) do
+    PendingChanges.create(%{data: job, params: %{"id" => job.id, "delete" => true}}, :delete, user)
```
Comment (doomspork) [inline lib/companies/jobs.ex]: Similar thought RE: `delete => true`

File: lib/companies/pending_changes.ex [hunk 7]
```diff
@@ -9,32 +9,15 @@ defmodule Companies.PendingChanges do
   alias Companies.Schema.{Company, Job, Industry, PendingChange}
 
   def all do
-    query =
-      from p in PendingChange,
-        select: p,
-        where: p.approved == false
+    query = from p in PendingChange, where: p.approved == false
```
Comment (doomspork) [inline lib/companies/pending_changes.ex]: What was the thinking behind this change?

File: lib/companies/pending_changes.ex [hunk 8]
```diff
@@ -9,32 +9,15 @@ defmodule Companies.PendingChanges do
...
-         changeset <- changeset(module, changes),
-         {:ok, _changes} <- apply_changes(action, changeset) do
-      change
-      |> PendingChange.changeset(%{approved: true})
-      |> Repo.update()
-    else
-      nil -> {:error, "change not found"}
-      {:error, changeset} -> {:error, changeset}
-    end
+  def get!(id) do
```
Comment (doomspork) [inline lib/companies/pending_changes.ex]: Do we really need to define this? We already have `get/1` and this isn't going to widely consume. I think a smaller API makes testing things easier and reduces the number of potential code paths

File: lib/companies/pending_changes.ex [hunk 9]
```diff
@@ -47,62 +30,66 @@ defmodule Companies.PendingChanges do
     end
   end
 
-  defp apply_changes("create", changeset), do: Repo.insert(changeset)
-  defp apply_changes("update", changeset), do: Repo.update(changeset)
+  def create(changeset, action, user, note \\ "")
```
Comment (doomspork) [inline lib/companies/pending_changes.ex]: Why did you move these? They were in alphabetical order before.

File: lib/companies/pending_changes.ex [hunk 10]
```diff
@@ -47,62 +30,66 @@ defmodule Companies.PendingChanges do
...
 
-  defp changeset(module, changes) do
-    record =
-      if Map.has_key?(changes, "id") do
-        id = Map.get(changes, "id")
-        Repo.get(module, id)
-      else
-        struct(module, %{})
-      end
+  def act(change_id, "false") do
```
Comment (doomspork) [inline lib/companies/pending_changes.ex]: I feel like this `act/2` functions is entirely unnecessary. Why can't we pass in a boolean to the existing method?

File: lib/companies/pending_changes.ex [hunk 11]
```diff
@@ -47,62 +30,66 @@ defmodule Companies.PendingChanges do
...
+    PendingChange
+    |> Repo.get!(change_id)
+    |> Repo.delete!()
```
Comment (doomspork) [inline lib/companies/pending_changes.ex]: We're not deleting anything. We just mark it as rejected. We want the record.

File: lib/companies/pending_changes.ex [hunk 12]
```diff
@@ -138,9 +142,17 @@ defmodule Companies.PendingChanges do
...
   end
 
+  defp apply_changes("create", changeset), do: Repo.insert(changeset)
+  defp apply_changes("update", changeset), do: Repo.update(changeset)
+  defp apply_changes("delete", record), do: Repo.delete(record)
+
   defp resource_module("company"), do: Company
   defp resource_module("industry"), do: Industry
   defp resource_module("job"), do: Job
+  defp resource_module(_), do: {:error, :module_not_recognized}
```
Comment (doomspork) [inline lib/companies/pending_changes.ex]: This should never happen, we don't need to worry about it. What brought you here?

File: lib/companies/pending_changes.ex [hunk 13]
```diff
@@ -138,9 +142,17 @@ defmodule Companies.PendingChanges do
...
+
+  defp drop_relations(job = %Job{}), do: Map.drop(job, [:company])
```
Comment (doomspork) [inline lib/companies/pending_changes.ex]: I don't think we need these methods potentially but if we do, they can probably be consolidated into one.  `Map.drop/2` can handle keys that don't exist.

File: lib/companies/pending_changes.ex (hunk 8 above)
Comment (gemantzu) [reply lib/companies/pending_changes.ex]: get/1 is doing some other things. Granted, this should be a private method and not parted of the API.

File: lib/companies/pending_changes.ex (hunk 7 above)
Comment (gemantzu) [reply lib/companies/pending_changes.ex]: I found the select redundant, so I removed it.

File: lib/companies/pending_changes.ex (hunk 9 above)
Comment (gemantzu) [reply lib/companies/pending_changes.ex]: Sorry, I didn't notice that you had them in alphabetical order and I moved them around in a way that made sense to me (and, I thought it might be in the longs run). I moved the api methods on top and the private below, moving the easy to deal with methods on the far bottom.

File: lib/companies/pending_changes.ex (hunk 10 above)
Comment (gemantzu) [reply lib/companies/pending_changes.ex]: It is passing by the value from the buttons in the form (accept => true, reject => false)

File: lib/companies/pending_changes.ex (hunk 11 above)
Comment (gemantzu) [reply lib/companies/pending_changes.ex]: Oh ok, nice. I will fix this.

**Comment by doomspork** [comment]
#466 includes this code 👍
